nmap.cc

```diff
@@ -1091,21 +1093,26 @@ void parse_options(int argc, char **argv) {
 
           while (i--) {
             do {
-              o.decoys[o.numdecoys].s_addr = get_random_u32();
-            } while (ip_is_reserved(&o.decoys[o.numdecoys]));
+              ((struct sockaddr_in *)&o.decoys[o.numdecoys])->sin_addr.s_addr = get_random_u32();
+            } while (ip_is_reserved(&((struct sockaddr_in *)&o.decoys[o.numdecoys])->sin_addr));
             o.numdecoys++;
           }
         } else {
           if (o.numdecoys >= MAX_DECOYS - 1)
             fatal("You are only allowed %d decoys (if you need more redefine MAX_DECOYS in nmap.h)", MAX_DECOYS);
 
           /* Try to resolve it */
-          struct sockaddr_in decoytemp;
-          size_t decoytemplen = sizeof(struct sockaddr_in);
-          int rc = resolve(p, 0, (sockaddr_storage*)&decoytemp, &decoytemplen, AF_INET);
+          struct sockaddr_storage decoytemp;
+          size_t decoytemplen = sizeof(struct sockaddr_storage);
+          int rc;
+          if (delayed_options.af == AF_INET6){
+            rc = resolve(p, 0, (sockaddr_storage*)&decoytemp, &decoytemplen, AF_INET6);
```

Comment:
Since the parsing of options go sequentially it's problematic with this kind of implementation, specifically when command is like `nmap -D bfea::a2,face:babe:face::1 2426:db8:0:f101::1 -6` then `delayed_options.af = AF_INET` hence it will try to resolve it as IPv4 address. If we want to get rid of it then either we can put this whole thing somewhere else or we can try to make a guess by seeing string `p`, both of them don't appear good solution I hope something better is there.